### ISSUE TITLE
Rcpp is not reverse compatible

### ISSUE THREAD

**Issue by XiangdongGu** [RcppCore/Rcpp#871]
The newer versions of Rcpp (0.12.16 and 0.12.17) broke all my codes that uses Rcpp. I don't know why but once I reverse to 0.12.15 everything is working.

Just curious is that something we expect?

**Comment by eddelbuettel** [comment]
See the [rcpp-logs repo](https://github.com/RcppCore/rcpp-logs) -- we run THOUSANDS of tests before each release, and CRAN also tests incoming package against all know dependencies.

So, this is neither expected nor known.  It would be polite of you to actually _back up such a claim_ with a [minimally reproducible example[(https://stackoverflow.com/questions/5963269/how-to-make-a-great-r-reproducible-example) -- because otherwise it is just hearsay and nothing we can do anything about.

**Comment by XiangdongGu** [comment]
Thank you @eddelbuettel ! My code was complex, so I extract the problem part and wrap into a smaller code.
Here is the cpp file
`
#include <Rcpp.h>
using namespace Rcpp;

// [[Rcpp::export]]
double loglik(NumericVector y, NumericMatrix X, NumericVector beta) {
  int J = X.ncol(), n = X.nrow();
  double result = 0.0, temp;
  for (int i=0; i < n; i++) {
    temp = 0;
    for (int j=0; j < J; j++) {
      temp += X(i, j) * beta[j];
    }
    result += (y[i] - temp)*(y[i] - temp);
  }
  return result;  
}

// [[Rcpp::export]]
NumericVector gradlik(NumericVector y, NumericMatrix X, NumericVector beta) {
  int J = X.ncol(), n = X.nrow(), i, j;
  double temp;
  NumericVector result(J);
  for (i=0; i<n; i++) {
    temp = 0;
    for (j=0; j<J; j++) {
      temp += X(i, j) * beta[j];
    }
    temp += - y[i];
    for (j=0; j<J; j++) {
      result[j] += 2* X(i, j) * temp;
    }
  }
  return result;
}

// [[Rcpp::export]]
void test(NumericMatrix X, NumericVector y, Function fitfun) {
  RNGScope scope;
  int J = X.ncol(), i;
  NumericVector par(J);
  for (i=0; i<J; i++) par[i] = 0;
  for (int i=0; i<10; i++) {
    Rprintf("%8f\n", log(R::runif(0.0, 1.0)));
    par = fitfun(par, y, X);
  }
}
`
Here is the R script

`
source("rcpp_new/.Rprofile", chdir = TRUE)
#source("rcpp_old/.Rprofile", chdir = TRUE)
library(Rcpp)
sourceCpp("test.cpp")

# Generate Test Data-----------------------------------------------------------
set.seed(123)
x <- rnorm(10)
y <- x + rnorm(10, sd = 0.1)
X <- cbind(1, x)

# Optimization function--------------------------------------------------------
fitfun <- function(parm, y, X) {
  q <- optim(parm, loglik, gradlik, method="BFGS", X = X, y = y)
  q$par
}

# New version Rcpp: the random number never change
# Old version Rcpp: the random number will change
test(X, y, fitfun)
`

My project code was to do some bayesian analysis with optimization routine (e.g. optimize quadratic function as an illustration here). So the random number generation is essential here. In old version of Rcpp (< 0.12.15), the random function will generate different values, however in newer version of Rcpp, it always gives same random number. This problem only happens when I put random function together with optimization function in the same cpp function as shown in test function in the code.

I am not familiar C++, so maybe my usage of C++ is not standard, but it would be helpful to understand what is causing this so I can modify my code to avoid any future failure.

Thanks for the help!

**Comment by eddelbuettel** [comment]
BTW `RNGScope scope;` should not be needed; Rcpp adds that for you already.  But look into the new `SuspendRNGSynchronizationScope` and how RcppDE uses it.

**Comment by XiangdongGu** [comment]
Thank you for @eddelbuettel for the clarification! I have been using old versions and waiting for new release for the fix. Recently, I have updated my R to version 3.5.1, and reinstalled all the packages including Rcpp (0.12.19). Unfortunately, the issues persist that it always gives same random number.  I was trying to downgrade to older version, e.g. 0.12.13, it throws error of 

```
In file included from Date.cpp:31:
In file included from ../inst/include/Rcpp.h:77:
../inst/include/Rcpp/Rmath.h:222:55: error: no member named 'Rf_pythag' in the global namespace; did you mean 'pythag'?
    inline double pythag(double a, double b)    { return ::Rf_pythag(a, b); }
                                                         ^~~~~~~~~~~
                                                         pythag
../inst/include/Rcpp/Rmath.h:222:19: note: 'pythag' declared here
    inline double pythag(double a, double b)    { return ::Rf_pythag(a, b); }
                  ^
1 error generated.
make: *** [Date.o] Error 1
ERROR: compilation failed for package ‘Rcpp’
* removing ‘/Volumes/Bayesian-64G/icensmis_packrat/packrat/lib/x86_64-apple-darwin15.6.0/3.5.1/Rcpp’
```

Really appreciate your help!

**Comment by eddelbuettel** [comment]
That error comes from 

- mixing an OLD version of Rcpp (which still uses `Rf_pythag`), here 0.12.13
- with a new verion of R (which no longer has it).  

That is your error in mixing, and your problem, caused the by the rare change in R -- they removed that symbol.  

CRAN and the rest of the world happily use matching (ie "current") versions.  We also just went through all of this in [issue #915](https://github.com/RcppCore/Rcpp/issues/915) which you could read in detail.

In short, you need to analyze the claim "Rcpp broke my code" and change your package to work with R 3.5.1.   Or, if you insist to work with your old packages, then you either

- must use an old R that was current at the time that Rcpp version was release
- update the old Rcpp you insist on using by not relying on `Rf_pythag`.

**Comment by XiangdongGu** [comment]
Thank you for the explanation and point me to the issue! For the random number issue, I tried to add  GetRNGstate() and PutRNGstate() for each call of random function, and it appears to be working.

It is just a little confusing to me that if my input function (fitfun in my above example) is regular function without involving optim, then the random state is changing as expected. But  if the function involves optim, then the random number is not moving. This seems to be happening only for newer Rcpp versions.

But I think GetRNGstate() and PutRNGstate() solve my problem. I can use all new versions of R and Rcpp. Thanks for the great package and help!

**Comment by XiangdongGu** [comment]
Hi @eddelbuettel,

After a little more investigation of my issue, I found that in the following case the RNG state is not changing: 1. the cpp function has Function type argument 2. the passed function called a cpp function.

I am illustrating a simple example as following

```
library(Rcpp)

sourceCpp(
  code  = '
  #include <Rcpp.h>
  using namespace Rcpp;

  // [[Rcpp::export]]
  double f(double x) {
    return x;
  }
  
  // [[Rcpp::export]]
  void test(Function myfun) {
    NumericVector y;
    for (int i = 0; i < 10; i++) {
       Rprintf("%8f\\n", runif(1)[0]);
       y = myfun(1);
    }
  }
  '
)

mfun <- function(x) f(1)

test(mfun)
```

In this example, mfun is R function and it calls a function defined in cpp. mfun is passed to test. In this case, the random number generated is no long changing. If I pass a function that does not call cpp function such as rnorm, then everything is fine.

I am not sure if this is known, but I would appreciate if you can give some suggestion how to best handle this.

Thanks.

**Comment by eddelbuettel** [comment]
I see.  You are calling back and forth between R and C++ which is both a little unusual and rare.  We did make changes in that area because we concluded that jumping back and forth like this is not the situation for which the RNG state preservations was designed.

This was discussed here in older issue tickets so please _do_ read the succint [release notes in NEWS.Rd](https://cloud.r-project.org/web/packages/Rcpp/news.html) where you will see that this was an area of work in release 0.12.18, 0.12.17 and 0.12.16 -- and each entry gives you a link to the discussion in the corresponding issue ticket.  We do think the change is for the better.  

In particular, see the discussion in #862 which should allow you to obtain the prior behaviour.

**Comment by eddelbuettel** [comment]
Yes, it is common practice to use _current R versions_ (which, by the way, is now R 3.6.1) with _current CRAN package versions_ which is after all how they are tested and quality-controlled at CRAN.  

So I recommend not only to use the current Rcpp version, but also the current R version.  Note that changing R versions also on occassion requires reinstallation of packages -- as documented in the release notes for R.

**Comment by byapparov** [comment]
I don't agree with the latest version recommendation.

Each team is free to decide when and why to upgrade. This is why packrat is so important.

**Comment by eddelbuettel** [comment]
This is not a useful discussion for Rcpp, so I am going to lock this.   I am glad Rcpp works for you, we put a lot of effort into it.
